File: Metro C#/Metro/Class/Subway.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Data;
using System.Windows.Media;
using System.Windows.Shapes;

namespace Metro
{
    public class Subway
    {
        public static readonly TextBox stationId = (TextBox)Application.Current.MainWindow.FindName("stationID");
        public static readonly Canvas canvas = (Canvas)Application.Current.MainWindow.FindName("canvas");
        public static readonly Canvas canvasPath = (Canvas)Application.Current.MainWindow.FindName("canvasPath");
        public static readonly TextBlock path = (TextBlock)Application.Current.MainWindow.FindName("path");
        public static readonly Button start = (Button)Application.Current.MainWindow.FindName("start");
        public static readonly ComboBox map = (ComboBox)Application.Current.MainWindow.FindName("map");
        public static readonly StackPanel load = (StackPanel)Application.Current.MainWindow.FindName("load");

        /// <summary>
        /// линии
        /// </summary>
        private List<Line> lines;
        /// <summary>
        /// все станции
        /// </summary>
        private List<Station> allStations;
        /// <summary>
        /// пересадки
        /// </summary>
        private List<Transfer> transfers;

        public Subway()
        {
            start.Click += Submit;

            canvasPath.Children.Add(Station.startPoint);
        }

        private async void Submit(object sender, RoutedEventArgs e)
        {
            start.IsEnabled = false;
            canvas.IsEnabled = false;
            map.IsEnabled = false;

            ClearCanvas();

            Station station = allStations.Find(s => s.ID == stationId.Text);

            ScrollViewer scrollPath = (ScrollViewer)path.Parent;

            load.Visibility = Visibility.Visible;
            scrollPath.Visibility = Visibility.Collapsed;
            List<Station> fullPath = await Task.Run(() => VisitAllLines(station));
            load.Visibility = Visibility.Collapsed;
            scrollPath.Visibility = Visibility.Visible;

            DrawPath(fullPath);

            List<string> outPath = fullPath.ConvertAll(a => a.ID);
            path.Text = string.Join(" --> ", outPath);

            start.IsEnabled = true;
            canvas.IsEnabled = true;
            map.IsEnabled = true;
        }

        /// <summary>
        /// создание схемы метро на основе входных данных
        /// </summary>
        /// <param name="inputData"></param>
        public void Bild(((SolidColorBrush color, (int x, int y)[] stations)[] lines, (string st1, string st2)[] transfers) inputData)
        {
            lines = new();
            allStations = new();
            transfers = new();

            Create(inputData);
            Draw(inputData);
        }

        private void Create(((SolidColorBrush color, (int x, int y)[] stations)[] lines, (string st1, string st2)[] transfers) inputData)
        {
            Line.NewID = 0;

            foreach (var line in inputData.lines)
            {
                Station.NewID = 0;

                Station prev = null;
                int i = 0;

                Line newLine = new Line(line.color);
                lines.Add(newLine);

                foreach (var stationPos in line.stations)
                {
                    newLine.Points.Add(new Point(stationPos.x, stationPos.y));

                    Station newStation = new Station(stationPos, newLine);

                    allStations.Add(newStation);
                    newLine.Stations.Add(newStation);

                    if (i++ != 0)
                    {
                        float distance = (float)Math.Ceiling(Math.Sqrt(Math.Pow(stationPos.x - prev.Position.x, 2) + Math.Pow(stationPos.y - prev.Position.y, 2)));

                        newStation.PrevStation = new(distance, prev);
                        prev.NextStation = new(distance, newStation);
                    }
                    prev = newStation;
                }
            }

            foreach (var transfer in inputData.transfers)
            {
                Station st1 = allStations.Find(station => station.ID == transfer.st1.ToUpper());
                Station st2 = allStations.Find(station => station.ID == transfer.st2.ToUpper());

                int x1 = st1 != null ? st1.Position.x : 0;
                int y1 = st1 != null ? st1.Position.y : 0;

                int x2 = st2 != null ? st2.Position.x : 0;
                int y2 = st2 != null ? st2.Position.y : 0;

                float distance = (float)Math.Ceiling(Math.Sqrt(Math.Pow(x1 - x2, 2) + Math.Pow(y1 - y2, 2)));

                Transfer newLink = new Transfer(distance, (st1, st2));

                if (st1 != null) st1.AddTransfer = newLink;
                if (st2 != null) st2.AddTransfer = newLink;

                transfers.Add(newLink);
            }
        }

        private void Draw(((SolidColorBrush color, (int x, int y)[] stations)[] lines, (string st1, string st2)[] links) inputData)
        {
            ClearCanvas();

            canvas.Children.Clear();

            foreach (var line in lines)
            {
                line.Draw();
            }

            foreach (var station in allStations)
            {
                station.Draw();
            }

            allStations[0].SelectStation();

            foreach (var transfer in transfers)
            {
                transfer.Draw();
            }
        }

        private static void ClearCanvas()
        {
            UIElement startPoint = canvasPath.Children[0];
            canvasPath.Children.Clear();
            canvasPath.Children.Add(startPoint);
        }


        /// <summary>
        /// поиск кротчайшего пути с посещением всех линий
        /// </summary>
        /// <param name="start">начальная станция</param>
        private List<Station> VisitAllLines(Station start)
        {
            lines.ForEach(line =>
            {
                line.Visited = false;
            });

            List<Station> path = new();
            List<Line> unVisitedLine = lines.ToList();
            Station station = start;            

            while (true)
            {
                // удаляет текущею линию
                unVisitedLine = unVisitedLine.FindAll(line => line.ID != station.Line.ID);

                // удаляет посещенные линии
                unVisitedLine = unVisitedLine.FindAll(line => line.Visited != true);

                // прекратить цикл если не осталось непосешенных линий
                if (unVisitedLine.Count == 0) break;

                FindShortWay(station);

                // ближайшие пересадки
                List<Station> nearestTransferStations = new();

                unVisitedLine.ForEach(line =>
                {
                    // только станции с пересадками
                    List<Station> transferStations = line.Stations.FindAll(
                      station => station.Transfers.Count > 0
                    );

                    // добавление ближайшей на линии
                    nearestTransferStations.AddRange(transferStations);
                });

                // сортировка по весу
                nearestTransferStations.Sort((a, b) => (int)(a.Path.Weight - b.Path.Weight));

                // следующей становится самая ближайшая
                station = nearestTransferStations[0];

                // 
                List<Station> newPath = station.Path.Full;

                newPath.ForEach(station =>
                {
                    station.Line.Visited = true;
                });

                path.AddRange(newPath);
            }

            path.Add(station);

            return path;
        }

        /// <summary>
        /// поиск кротчайшего пути от текущей станции до каждой
        /// </summary>
        /// <param name="current"></param>
        private void FindShortWay(Station current)
        {
            // сброс просчитанных путей
            allStations.ForEach(station =>
            {
                station.Path.Weight = int.MaxValue;
                station.Path.Full = new List<Station>();
            });

            // устанавливает начальной станции вес в 0
            current.Path.Weight = 0;

            Queue<Station> queue = new();

            while (true)
            {
                List<NeighboringStation> possibleWays = new();
                NeighboringStation prev = current.PrevStation;
                NeighboringStation next = current.NextStation;
                List<NeighboringStation> transfers = current.Transfers;

                if (prev != null) possibleWays.Add(prev);

                if (next != null) possibleWays.Add(next);

                if (transfers != null) possibleWays.AddRange(transfers);

                Debug.WriteLine($"\ncurent {current.ID}, W= {current.Path.Weight}");

                List<string> a = possibleWays.ConvertAll(a => $"ID= {a.Station.ID}, W= {a.Station.Path.Weight}");
                Debug.WriteLine($"possibleWays {possibleWays.Count}= {string.Join("; ", a)}");

                if (possibleWays.Count > 0)
                {
                    possibleWays.Sort((a, b) => (int)(a.Dist - b.Dist));

                    possibleWays.ForEach(way =>
                    {
                        float newWeight = current.Path.Weight + way.Dist;
                        float oldWeight = way.Station.Path.Weight;

                        Debug.WriteLine($"way {way.Station.ID} new W= {newWeight}, old W= {oldWeight}");

                        // добавление в очередь любой станции если путь стал короче
                        if (newWeight < oldWeight)
                        {
                            way.Station.Path.Weight = newWeight;

                            List<Station> newPath = current.Path.Full.ToList();
                            newPath.Add(current);

                            way.Station.Path.Full = newPath;

                            queue.Enqueue(way.Station);
                            Debug.WriteLine($"{way.Station.ID} add to queue");
                        }
                    });
                }

                if (queue.Count > 0)
                {
                    current = queue.Dequeue();
                }
                else
                {
                    break;
                }
            }

            Debug.WriteLine($"\nallStations");
            allStations.ForEach(s =>
            {
                List<string> d = s.Path.Full.ConvertAll(p => $"ID= {p.ID}, W= {p.Path.Weight}");
                Debug.WriteLine($"\n{s.ID} Path {s.Path.Full.Count}= {string.Join("; ", d)}");
            });
        }

        /// <summary>
        /// отрисовка пути
        /// </summary>
        private void DrawPath(List<Station> path)
        {
            PointCollection points = new();

            path.ForEach(station =>
            {
                points.Add(new Point(station.Position.x, station.Position.y));
            });

            Polyline polyline = new()
            {
                Points = points,
                Stroke = new SolidColorBrush(Colors.Black),
                StrokeThickness = 8,
                StrokeEndLineCap = PenLineCap.Triangle,
                StrokeStartLineCap = PenLineCap.Round,
                StrokeLineJoin = PenLineJoin.Round,
                IsHitTestVisible = false
            };

            Panel.SetZIndex(polyline, 2);

            canvasPath.Children.Add(polyline);
        }
    }
}

File: Метро JS/js/Subway.js
// метро
class Subway {
  #lines = [];
  #allStations = [];
  #links = [];
  #ctx;

  /* get lines() {
    return this.#lines;
  } */

  constructor(ctx, lines = [], links = []) {
    this.#ctx = ctx;
    this.CreateLine(lines);

    console.log('links', this.#links);
    console.log('lines', this.#lines);
    console.log('allStations', this.#allStations);

    this.CreateLink(links);

    this.DrawSubway();

    let startID = window.location.search.replace(/.+map=(.+)&id=(.+)/, '$2');
    if (startID == '') startID = 'l0_s1';
    document.getElementById('station').value = startID;

    let start = this.#allStations.find((station) => station.Id == startID);
    this.VisitAllLines(start);
    // this.FindShortWay(start);
  }

  CreateLine(lines) {
    lines.forEach((line) => {
      let prev;
      let i = 0;

      let color = line.line.color;
      let name = line.line.name != undefined ? line.line.name : '';
      let newline = new Line(color, name);

      let newStationsA = [];
      line.stations.forEach((station) => {
        let name = station.name != undefined ? station.name : '';
        let newStation = new Station(
          { x: station.x, y: station.y },
          newline,
          name
        );

        if (i++ != 0) {
          let distance = Math.ceil(
            Math.sqrt(
              Math.pow(station.x - prev.Coordinate.x, 2) +
                Math.pow(station.y - prev.Coordinate.y, 2)
            )
          );
          // console.log(distance);
          newStation.PrevStation = { dist: distance, station: prev };
          prev.NextStation = { dist: distance, station: newStation };
        }
        prev = newStation;

        newStationsA.push(newStation);
      });

      newline.Stations = newStationsA;
      this.#lines.push(newline);
      this.#allStations = this.#allStations.concat(newStationsA);
    });
  }

  CreateLink(links) {
    console.log(links);
    links.forEach((link) => {
      let station1 = this.#allStations.find(
        (station) => station.Id == link.st1
      );
      let station2 = this.#allStations.find(
        (station) => station.Id == link.st2
      );

      let x1 = station1.Coordinate.x;
      let y1 = station1.Coordinate.y;
      let x2 = station2.Coordinate.x;
      let y2 = station2.Coordinate.y;

      let distance = Math.ceil(
        Math.sqrt(Math.pow(x1 - x2, 2) + Math.pow(y1 - y2, 2))
      );

      let newLink = new Link(distance, [station1, station2]);
      station1.AddLink = newLink;
      station2.AddLink = newLink;
      this.#links.push(newLink);
    });
  }

  DrawSubway() {
    this.#ctx.clearRect(0, 0, 1000, 1000);
    this.#lines.forEach((line) => {
      line.DrawLine(this.#ctx);
    });

    this.#links.forEach((link) => {
      link.DrawLink(this.#ctx);
    });

    this.#allStations.forEach((station) => {
      station.DrawStation(this.#ctx);
    });
  }

  VisitAllLines(start) {
    // поиск кротчайшего пути с посещением всех линий
    let path = [];
    let unVisitedLine = this.#lines;

    let station = start;

    let i = 0;
    while (true) {
      // удаляет текущею линию
      unVisitedLine = unVisitedLine.filter(
        (line) => line.Id != station.Line.Id
      );
      // удаляет посещенные линии
      unVisitedLine = unVisitedLine.filter((line) => line.Visited != true);

      if (unVisitedLine.length == 0) break;

      this.FindShortWay(station);

      console.log('unVisitedLine', unVisitedLine);

      // let singleLinkStation = [];
      // let multipleLinkStation = [];

      // ближайшие пересадки
      let nearestTransferStations = [];

      unVisitedLine.forEach((line) => {
        // только станции с пересадками
        let linkSt = line.Stations.filter(
          (station) => station.Links.length > 0
        );
        console.log(linkSt);

        // сортировка по весу
        // linkSt.sort((a, b) => a.Path.weight - b.Path.weight);

        // добавление ближайшей на линии
        nearestTransferStations = nearestTransferStations.concat(linkSt);

        /* if (linkSt.length == 1) {
          singleLinkStation = singleLinkStation.concat(linkSt);
        } else {
          multipleLinkStation = multipleLinkStation.concat(linkSt);
        } */
      });

      // сортировка по весу
      nearestTransferStations.sort((a, b) => a.Path.weight - b.Path.weight);

      // следующей становится самая ближайшая
      station = nearestTransferStations[0];
      console.log('nearestTransferStations', nearestTransferStations);

      /* singleLinkStation.sort((a, b) => a.Path.weight - b.Path.weight);
      console.log('singleLinkStation', singleLinkStation);

      multipleLinkStation.sort((a, b) => a.Path.weight - b.Path.weight);
      console.log('multipleLinkStation', multipleLinkStation); */

      /* if (singleLinkStation.length > 0) {
        station = singleLinkStation[0];
      } else {
        station = multipleLinkStation[0];
      } */

      let newPath = station.Path.path;
      newPath.forEach((station) => {
        station.Line.Visited = true;
      });

      path = path.concat(newPath);
    }
    path.push(station);
    console.log('path', path);
    this.DrawPath(path);

    path = path.map((a) => a.Id);
    console.log(path);
    document.getElementById('path').value = path.join(' --> ');
  }

  ResetPath() {
    // сброс просчитанных путей
    this.#allStations.forEach((station) => {
      station.Path = { /*visited: false,*/ weight: Infinity, path: [] };
    });
  }

  FindShortWay(current) {
    // поиск кротчайшего пути от текущей станции до каждой
    this.ResetPath();
    current.Path = { weight: 0 };

    // let unVisitedStations = this.#allStations;

    let queue = [];

    while (true) {
      // current.Path = { visited: true };

      // unVisitedStations = unVisitedStations.filter(
      //   (station) => station != current
      // );

      console.log('\n! current:', current.Id, current);
      // console.log('unVisitedStations', unVisitedStations);

      let possibleWays = [];
      let prev = current.PrevStation;
      let next = current.NextStation;
      let links = current.Links;

      if (prev != undefined) {
        possibleWays.push(prev);
      }
      if (next != undefined) {
        possibleWays.push(next);
      }
      if (links != undefined) {
        possibleWays = possibleWays.concat(links);
      }
      console.log('possibleWays:', possibleWays);

      /* possibleMove = possibleMove.filter(
        (move) => move.station.Path.visited != true
      ); */

      if (possibleWays.length > 0) {
        possibleWays.sort((a, b) => a.dist - b.dist);

        // current.DrawMark(this.#ctx, 0);

        let waysId = [];

        possibleWays.forEach((way) => {
          waysId.push(way.station.Id);

          let newWeight = current.Path.weight + way.dist;
          let weight = way.station.Path.weight;

          // добавление в очередь любой станции если путь стал короче
          if (newWeight < weight) {
            way.station.Path = {
              weight: newWeight,
              path: current.Path.path.concat(current),
            };
            queue.push(way.station);
            console.log('add to queue:', way.station);
          }
        });
        // console.log('possibleMove:', movesId.join(', '), possibleMove);
      }

      console.log('queue:', ...queue);

      if (queue.length > 0) {
        current = queue.shift();
      } else {
        break;
      }

      // console.log('stack', [...stack]);
    }

    let stations_path = [];
    this.#allStations.forEach((station) => {
      let id = { id: station.Id };
      // id += station.Path;
      stations_path.push(Object.assign(id, station.Path));
    });
    console.log('stations_path:', stations_path);
  }

  DrawPath(path, color = 'black', width = 6) {
    let ctx = this.#ctx;
    ctx.beginPath();
    ctx.strokeStyle = color;
    ctx.lineWidth = width;

    let i = 0;
    path.forEach((station) => {
      let x = station.Coordinate.x;
      let y = station.Coordinate.y;

      if (i == 0) {
        ctx.moveTo(x, y);
        i = 1;
      } else {
        ctx.lineTo(x, y);
      }
    });
    ctx.stroke();
  }
}
